### ISSUE TITLE
TransactionManagers.create() crashes if it can't talk to Cassandra within a few seconds

### ISSUE THREAD

**Issue by jeremyk-91** [palantir/atlasdb#1148]
An internal product ran into flakes in their ETE tests when trying to start up AtlasDB, because the TransactionManagers factory failed during the health checks (possibly because Cassandra was slow to start). I suspect this is related to #1028 - the offending part of the stacktrace in that issue also appeared in the stacktrace that I saw.

The current approach is to ping each server twice - once with describe_version() and, if that is successful, validating its partitioner (see `CassandraClientPool#runOneTimeStartupChecks()` for more detail). Note that as long as at least 1 host responds we can go ahead, but their ETE tests only have 1 host which may perhaps have contributed to making these issues more visible.

A redacted version of the relevant stacktraces we experienced are below:

```java
Exception in thread "main" java.lang.RuntimeException: Performing routine startup checks, determined that the following hosts are unreachable for the following reasons: 
	Host: <HOST> was marked unreachable via exception: com.palantir.atlasdb.keyvalue.cassandra.CassandraClientFactory$ClientCreationFailedException: Failed to construct client for <CLIENT> over SSL

	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPool.runOneTimeStartupChecks(CassandraClientPool.java:362)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueService.init(CassandraKeyValueService.java:213)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueService.create(CassandraKeyValueService.java:177)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueService.create(CassandraKeyValueService.java:163)
	at com.palantir.atlasdb.cassandra.CassandraAtlasDbFactory.createKv(CassandraAtlasDbFactory.java:48)
	at com.palantir.atlasdb.cassandra.CassandraAtlasDbFactory.createRawKeyValueService(CassandraAtlasDbFactory.java:42)
	at com.palantir.atlasdb.factory.ServiceDiscoveringAtlasSupplier.lambda$new$7(ServiceDiscoveringAtlasSupplier.java:47)
	at com.google.common.base.Suppliers$MemoizingSupplier.get(Suppliers.java:120)
	at com.palantir.atlasdb.factory.ServiceDiscoveringAtlasSupplier.getKeyValueService(ServiceDiscoveringAtlasSupplier.java:52)
	at com.palantir.atlasdb.factory.TransactionManagers.create(TransactionManagers.java:122)
	at com.palantir.atlasdb.factory.TransactionManagers.create(TransactionManagers.java:107)
	at com.palantir.atlasdb.factory.TransactionManagers.create(TransactionManagers.java:95)
	<rest of stack trace removed>
```

I can see two possible ways of moving forward:
* introducing some retrying / leniency in the health checks. This is easy, but the other solution feels more complete.
* lazy initialisation or at least not immediately throwing if we can't yet talk to the KVS in the transaction manager. I'm not sure if this has other implications; if we try to use the KVS pretty shortly after we may run into issues anyway.

**Comment by tpetracca** [comment]
@jeremyk-91, we mapped out some of how we thought this should work in #930

**Comment by jeremyk-91** [comment]
@tpetracca Thanks for the link to the other ticket. I've read through the discussion over there and am in agreement with how to proceed (i.e. maintain 2 paths, 1 fail-fast (current behaviour) and 1 that tolerates it, but throws immediately if Cassandra is unavailable) as well. I'm not sure how frequently this issue is occurring across internal clients, but will have a look at PDS-45417 and raise it with the team tomorrow.

**Comment by tpetracca** [comment]
So I'd classify this as one of those things that isn't a P0 for anyone single customer, but is a pain in the ass for EVERY customer and thus is probably worth doing.  There is a point in the future where a specific consumer product is going to need and be blocked by this as well (we're just not there yet).  Happy to talk further offline.

**Comment by hsaraogi** [comment]
Duplicate of #930.
